### ISSUE TITLE
WC_Admin_Settings::get_option() is forcing array values to be of 'string' due to array map stripslashes.

### ISSUE THREAD

**Issue by junixdev** [woocommerce/woocommerce#18645]
This only happens for "multiselect" fields that saves a list of integer option values. The one causing it to be string is the code in WooCommerce in this line https://github.com/woocommerce/woocommerce/blob/master/includes/admin/class-wc-admin-settings.php#L181

Also in this line, https://github.com/woocommerce/woocommerce/blob/master/includes/admin/class-wc-admin-settings.php#L381 WooCommerce sets a multiple option as selected via in_array as **strict**, so it is only true when the value types are the same.

**Possible solutions:**
- add a **intval array_map** after the stripslashes array_map.
- undo the strict on **in_array** on line 381.

## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Create a new "multiselect" setting field that uses integer as values. below is the code for adding a sample one in general settings:
```
add_filter( 'woocommerce_general_settings' , 'wcissue_custom_multiselect_setting' );
function wcissue_custom_multiselect_setting( $settings ) {

	$settings[] = array( 'name' => 'Custom Settings', 'type' => 'title', 'desc' => '', 'id' => 'wcissue_custom_multiselect_setting' );

    $settings[] = array(
        'title'   => 'Sample Custom Multiselect',
        'id'      => 'wcissue_custom_multiselect_field',
        'type'    => 'multiselect',
		'class'   => 'wc-enhanced-select',
        'options' => array(
            10 => 'option 10',
            20 => 'option 20',
            30 => 'option 30',
            40 => 'option 40',
            50 => 'option 50',
            60 => 'option 60'
        ),
        'placeholder' => 'Select options'
    );

	$settings[] = array( 'type' => 'sectionend', 'id' => 'wcissue_custom_multiselect_setting');

	return $settings;
}
```
Please note that this is just an example field. In actuality, what we are saving/using is the product ID.

2. Go to WC general settings, select some options and hit save.
3. **Issue:** selected options are not showing up.

## Expected/actual behavior

When I follow those steps, I see... options saved are not selected.

I was expecting to see... options saved to be selected.

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
```

### WordPress Environment ###

Home URL: http://www.wwlc.test
Site URL: http://www.wwlc.test
WC Version: 3.3.0
Log Directory Writable: ✔
WP Version: 4.9.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: Apache/2.2.32 (Unix) mod_wsgi/3.5 Python/2.7.13 PHP/7.1.8 mod_ssl/2.2.32 OpenSSL/1.0.2j DAV/2 mod_fastcgi/2.4.6 mod_perl/2.0.9 Perl/v5.24.0
PHP Version: 7.1.8
PHP Post Max Size: 32 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.52.1
OpenSSL/1.0.2j

SUHOSIN Installed: –
MySQL Version: 5.6.35
Max Upload Size: 32 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ❌ wp_remote_post() failed. Contact your hosting provider. cURL error 28: Operation timed out after 10000 milliseconds with 0 bytes received
Remote Get: ❌ wp_remote_get() failed. Contact your hosting provider. 200

### Database ###

WC Database Version: 3.2.6
WC Database Prefix: wp_
MaxMind GeoIP Database: ❌ The MaxMind GeoIP Database does not exist - Geolocation will not function. You can download and install it manually from http://dev.maxmind.com/geoip/legacy/geolite/ to the path: . Scroll down to "Downloads" and download the "Binary / gzip" file next to "GeoLite Country". Please remember to uncompress GeoIP.dat.gz and upload the GeoIP.dat file only.
Total Database Size: 14.81MB
Database Data Size: 8.41MB
Database Index Size: 6.40MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB
wp_comments: Data: 0.02MB + Index: 0.09MB
wp_icl_content_status: Data: 0.02MB + Index: 0.02MB
wp_icl_core_status: Data: 0.02MB + Index: 0.02MB
wp_icl_flags: Data: 0.02MB + Index: 0.02MB
wp_icl_languages: Data: 0.02MB + Index: 0.03MB
wp_icl_languages_translations: Data: 0.19MB + Index: 0.11MB
wp_icl_locale_map: Data: 0.02MB + Index: 0.00MB
wp_icl_message_status: Data: 0.02MB + Index: 0.03MB
wp_icl_mo_files_domains: Data: 0.02MB + Index: 0.02MB
wp_icl_node: Data: 0.02MB + Index: 0.00MB
wp_icl_reminders: Data: 0.02MB + Index: 0.00MB
wp_icl_strings: Data: 2.52MB + Index: 3.48MB
wp_icl_string_packages: Data: 0.02MB + Index: 0.00MB
wp_icl_string_pages: Data: 0.08MB + Index: 0.06MB
wp_icl_string_positions: Data: 0.02MB + Index: 0.02MB
wp_icl_string_status: Data: 0.02MB + Index: 0.02MB
wp_icl_string_translations: Data: 2.52MB + Index: 1.52MB
wp_icl_string_urls: Data: 0.02MB + Index: 0.02MB
wp_icl_translate: Data: 0.02MB + Index: 0.02MB
wp_icl_translate_job: Data: 0.02MB + Index: 0.02MB
wp_icl_translations: Data: 0.02MB + Index: 0.06MB
wp_icl_translation_batches: Data: 0.02MB + Index: 0.00MB
wp_icl_translation_status: Data: 0.02MB + Index: 0.02MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_options: Data: 2.06MB + Index: 0.05MB
wp_postmeta: Data: 0.09MB + Index: 0.06MB
wp_posts: Data: 0.06MB + Index: 0.06MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.02MB + Index: 0.03MB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wp_usermeta: Data: 0.05MB + Index: 0.03MB
wp_users: Data: 0.02MB + Index: 0.05MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB

### Post Type Counts ###

attachment: 16
nav_menu_item: 6
page: 20
post: 1
product: 15
product_variation: 4
revision: 23
shop_order: 1

### Security ###

Secure connection (HTTPS): ❌Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (8) ###

WPML Multilingual CMS: by OnTheGoSystems – 3.8.3
WooCommerce Wholesale Lead Capture: by Rymera Web Co – 1.7.0 – Not tested with the active version of WooCommerce
WooCommerce Wholesale Order Form: by Rymera Web Co – 1.8.0 – Not tested with the active version of WooCommerce
WooCommerce Wholesale Prices Premium: by Rymera Web Co – 1.16.0 – Not tested with the active version of WooCommerce
WooCommerce Wholesale Prices: by Rymera Web Co – 1.6.0 – Not tested with the active version of WooCommerce
WooCommerce: by Automattic – 3.3.0-rc.2
WPML String Translation: by OnTheGoSystems – 2.6.2
WPML Translation Management: by OnTheGoSystems – 2.4.2

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: PHP (₱)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My account: #7 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 2.2.7
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –


```
</details>

**Comment by shivapoudel** [comment]
Duplicate of #18102 and has been patched here #18186 :)

**Comment by jkohlbach** [comment]
@mikejolley this is happening in WC3.3 RC2

**Comment by jkohlbach** [comment]
@shivapoudel thanks, are you getting this still on 3.3 RC2? We are.

**Comment by shivapoudel** [comment]
@kloon @mikejolley Confirmed. Saving select settings seems fine but is not for multiselect :)

**Comment by mikejolley** [comment]
HTML still has no concept of integers so you really shouldn't be using them. We'll look for a workaround but just swap to strings..

**Comment by jkohlbach** [comment]
@mikejolley Sorry, I don't think that's good enough, from what I understand you can validly supply integers to a select <option>'s values in HTML and it's commonplace to do so.

WooCommerce sets a select box with "multiple" option as selected via in_array as strict, so it is only true when the value types are the same.

It's a very simple solution and it's also been working in the past so this is a regression in 3.3.

If we were to follow your suggestion we'll basically be losing all the custom settings admins have in this setting which is a horrible experience for users.

Also it was fixed for single selects, so why not for multiselect? It was already patched in #18102 for single selects.
